app/db.py

```diff
@@ -0,0 +1,501 @@
+            url = "https://www.ncbi.nlm.nih.gov/pmc/utils/idconv/v1.0/?ids="
+            for row in list_dois_tmp:
+                url += quote(row.doi) + ","
+                dois_requested[row.doi] = row
+            url = url[:-1]
+
+            # send request to NCBI API
+            resp = request_ncbi_api(url, ncbi_tool, ncbi_email)
+            resp_data = resp.json()
+            for doi, row in dois_requested.items():
+                request_dict = {
+                    "doi": doi,
+                    "request_url": url,
+                    "request_type": "ncbi",
+                    "response_content": dumps(resp_data),
+                    "response_status": resp.status_code,
+                }
+                db_requests_added.append(request_dict)
+                row.url_ncbi = True
+
+            if "records" in resp_data:
+                for rec in resp_data["records"]:
+                    # create PMC url
+                    if "pmcid" in rec:
+                        url_pmc = "https://ncbi.nlm.nih.gov/pmc/articles/PMC{0}/".format(
+                            quote(rec["pmcid"])
+                        )
+                        if url_pmc not in url_list + urls_added:
+                            url_dict = {
+                                "url": url_pmc,
+                                "doi": rec["doi"],
+                                "url_type": "pmc",
+                            }
+                            db_urls_added.append(url_dict)
+                            urls_added.append(url_pmc)
+                            num_urls_pmc_added += 1
+                            if rec["doi"] in dois_requested:
+                                dois_requested[rec["doi"]].url_pmc = True
+                    # create PM url
+                    if "pmid" in rec:
+                        url_pm = "https://www.ncbi.nlm.nih.gov/pubmed/{0}".format(
+                            rec["pmid"]
+                        )
+                        if url_pm not in url_list + urls_added:
+                            url_dict = {
+                                "url": url_pm,
+                                "doi": rec["doi"],
+                                "url_type": "pm",
+                            }
+                            db_urls_added.append(url_dict)
+                            urls_added.append(url_pm)
+                            num_urls_pm_added += 1
+                            if rec["doi"] in dois_requested:
+                                dois_requested[rec["doi"]].url_pm = True
+        create_entities(db, Request, db_requests_added)
+        create_entities(db, Url, db_urls_added)
+        num_requests_added += len(db_requests_added)
+    print('{0} "PM" URL\'s added to database.'.format(num_urls_pm_added))
+    print('{0} "PMC" URL\'s added to database.'.format(num_urls_pmc_added))
+    print("{0} Requests added to database.".format(num_requests_added))
+
+
+def create_unpaywall_urls() -> None:
```

Comment:
`flask unpaywall` isn't working for me currently.

```
sqlalchemy.exc.IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "urls_pkey"
DETAIL:  Key (url)=(https://doi.org/10.22230/src.2011v2n1a29) already exists.

[SQL: INSERT INTO urls (created_at, updated_at, url, doi, url_type) VALUES (%(created_at)s, %(updated_at)s, %(url)s, %(doi)s, %(url_type)s)]
```

I can provide the full traceback, but I think that it's mostly something about the CRUD calls?

Reply:
This happens, then the unique constrain of the attribute url in the url-table is violated. maybe you need to reset the table(s). It often happens, when you import the same data twice.